plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/N4JSDocument.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2017 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ui.editor;
+
+import java.lang.reflect.Field;
+import java.util.concurrent.locks.Lock;
+
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.ui.editor.model.DocumentTokenSource;
+import org.eclipse.xtext.ui.editor.model.XtextDocument;
+import org.eclipse.xtext.ui.editor.model.edit.ITextEditComposer;
+import org.eclipse.xtext.util.concurrent.IUnitOfWork;
+import org.eclipse.xtext.xbase.lib.util.ReflectExtensions;
+
+import com.google.inject.Inject;
+
+/**
+ * Minor customization of {@link XtextDocument} in order to make {@link Lock#tryLock()} available to clients of Xtext
+ * documents.
+ */
+public class N4JSDocument extends XtextDocument {
+
+	/** Copy of private field with same name from super class for local use. */
+	private final N4JSDocumentLocker stateAccess;
+
+	/**
+	 * Creates a new instance, delegates to {@link XtextDocument#XtextDocument(DocumentTokenSource, ITextEditComposer)}.
+	 */
+	@Inject
+	public N4JSDocument(DocumentTokenSource tokenSource, ITextEditComposer composer) {
+		super(tokenSource, composer);
+		stateAccess = (N4JSDocumentLocker) readField(XtextDocument.class, "stateAccess", this);
+	}
+
+	/**
+	 * Like {@link #readOnly(IUnitOfWork)}, but won't block if a read-lock cannot be acquired at this time; instead, the
+	 * given default value will be returned in this case.
+	 */
+	public <T> T tryReadOnly(IUnitOfWork<T, XtextResource> work, T defaultValue) {
+		return stateAccess.tryReadOnly(work, defaultValue);
+	}
+
+	@Override
+	protected XtextDocumentLocker createDocumentLocker() {
+		return new N4JSDocumentLocker();
+	}
+
+	/** See {@link N4JSDocument}. */
+	protected class N4JSDocumentLocker extends XtextDocumentLocker {
+
+		/** Copy of private field with same name from super class for local use. */
+		private final Lock readLock;
+
+		/** Creates a new instance. */
+		public N4JSDocumentLocker() {
+			super();
+			readLock = (Lock) readField(XtextDocumentLocker.class, "readLock", this);
+		}
+
+		/** See {@link N4JSDocument#tryReadOnly(IUnitOfWork, Object)}. */
+		public <T> T tryReadOnly(IUnitOfWork<T, XtextResource> work, T defaultValue) {
```

Comment:
Can we somehow use a result value that carries the information whether the lock was obtained or not?

Reply:
I was considering that, also because it would be more in line with the API of `Lock#tryLock()` which returns a boolean.

However, it would make using this method much more unwieldy and in all our current use cases the pattern with the default value works fine. Did not want to add another tryReadOnly() method that is not used yet (in our code).

But if this is added to Xtext, there should definitely be a method as you propose (probably in addition to the more convenient method with the default value).